src/main/host/thread_ptrace.c

```diff
@@ -406,9 +438,136 @@ static void threadptrace_flushPtrs(Thread* base) {
     _threadptrace_flushPtrs(thread);
 }
 
+static void _threadptrace_doAttach(ThreadPtrace* thread) {
+    utility_assert(thread->childState == THREAD_PTRACE_CHILD_STATE_SYSCALL);
+
+    debug("thread %i attaching to child %i", thread->base.threadID, (int)thread->base.nativePid);
+    if (ptrace(PTRACE_ATTACH, thread->base.nativePid, 0, 0) < 0) {
+        error("ptrace: %s", g_strerror(errno));
+        abort();
+    }
+    int wstatus;
+    if (waitpid(thread->base.nativePid, &wstatus, 0) < 0) {
+        error("waitpid: %s", g_strerror(errno));
+        abort();
+    }
+    StopReason reason = _getStopReason(wstatus);
+    utility_assert(reason.type == STOPREASON_SIGNAL && reason.signal.signal == SIGSTOP);
+
+    if (ptrace(PTRACE_SETOPTIONS, thread->base.nativePid, 0, THREADPTRACE_PTRACE_OPTIONS) < 0) {
+        error("ptrace: %s", strerror(errno));
+        return;
+    }
+
+#if DEBUG
+    // Check that regs are where we left them.
+    struct user_regs_struct regs;
+    if (ptrace(PTRACE_GETREGS, thread->base.nativePid, 0, &regs) < 0) {
+        error("ptrace: %s", g_strerror(errno));
+        abort();
+    }
+    utility_assert(!memcmp(&regs, &thread->syscall.regs, sizeof(regs)));
+#endif
+
+    // Should cause syscall we stopped at to be re-executed, putting us back in syscall state
+    while (reason.type == STOPREASON_SIGNAL && reason.signal.signal == SIGSTOP) {
+        if (ptrace(PTRACE_SYSEMU, thread->base.nativePid, 0, 0) < 0) {
+            error("ptrace: %s", g_strerror(errno));
+            abort();
+        }
+        if (waitpid(thread->base.nativePid, &wstatus, 0) < 0) {
+            error("waitpid: %s", g_strerror(errno));
+            abort();
+        }
+        reason = _getStopReason(wstatus);
+    }
+    if (reason.type != STOPREASON_SYSCALL) {
+        error("unexpected stop reason: %d", reason.type);
+        utility_assert(reason.type == STOPREASON_SYSCALL);
+    }
+
+    // Restore our saved instruction pointer to where it was before we detached - just after the
+    // syscall insn.
```

Comment:
What is "insn"? Instruction?

Reply:
Yup; expanded to 'instruction' for clarity